### PR TITLE
Chore: improve typings which support Typescript

### PR DESCRIPTION
**Pull request by ttkien**
# Motivation:
- Refactor code and configuration support better for typescript

# Changes
- Support write test in typescript
- Hero Theme:
- Add tests
- Component:
## Avatar
avatar

## Badge
badge

## Button
CleanShot 2021-07-10 at 18 52 21

## DateTimePicker
CleanShot 2021-07-10 at 18 56 36

## Error Screen 
CleanShot 2021-07-10 at 18 57 10

## MentionTextInput
CleanShot 2021-07-10 at 18 57 28

## Message
CleanShot 2021-07-10 at 18 57 47

## RichTextEditor
CleanShot 2021-07-10 at 18 58 04

## Text
CleanShot 2021-07-10 at 18 58 19
CleanShot 2021-07-10 at 18 58 26
CleanShot 2021-07-10 at 18 58 34
CleanShot 2021-07-10 at 18 58 46

## TextInput
CleanShot 2021-07-12 at 14 09 55

## TouchableTextInput
CleanShot 2021-07-12 at 14 10 08

## Container
CleanShot 2021-07-12 at 14 10 32

Demo

https://user-images.githubusercontent.com/1519521/125188825-7ce3b380-e25f-11eb-8f50-f48d56dce645.mp4

### CODE REVIEW

File: lib/src/bindings/Helpers.gen.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  ? Y
+  : N;
+declare const exactType: <T, U>(
+  draft: T & IfEquals<T, U>,
+  expected: U & IfEquals<T, U>,
+) => IfEquals<T, U>;
+
+//if Hero_Theme is not match to Legacy_Theme, the compile shows error
+exactType(Hero_Theme, Legacy_Theme); // error
+
```
Comment (ttkien) [inline lib/src/bindings/Helpers.gen.ts:24]: Verify Hero and Legacy Themes

File: lib/src/bindings/Events.re [hunk 2]
```diff
@@ -14,7 +14,7 @@ type listener = arguments => unit;
 
 [@bs.send] external emit: (t, eventName, arguments) => unit = "emit";
 
-[@bs.send] external setMaxListeners: (t, int) => t = "setMaxListeners";
+[@bs.send] external setMaxListeners: (t, int) => unit = "setMaxListeners";
```
Comment (ttkien) [inline lib/src/bindings/Events.re:17]: In bs-platform 8.0+, old code raised compilation error

File: lib/package.json [hunk 3]
```diff
@@ -9,17 +9,24 @@
     "watch": "bsb -make-world -w & webpack --config webpack.lib.js -w",
     "build": "bsb -clean-world -make-world && webpack --config webpack.lib.js && yarn ts:build",
     "pdf-editor:build": "webpack --config webpack.pdfEditorApp.js",
-    "test": "bsb -make-world && jest",
+    "test": "bsb -make-world && jest --runInBand",
```
Comment (ttkien) [inline lib/package.json:12]: Fix error: reach maximum memory on CircleCI

File: lib/src/components/Avatar.re [hunk 4]
```diff
@@ -12,7 +12,11 @@ type stylesBySize = {
   "image": Style.t,
 };
 
-let getStylesBySize: (option(string), Js.t('a)) => stylesBySize =
+[@genType]
+[@bs.deriving jsConverter]
+type size = [ | `small | `large | `medium];
```
Comment (toan2406) [inline lib/src/components/Avatar.re:17]: Are those polymorphic variants transpiled to string?

Comment (ttkien) [reply lib/src/components/Avatar.re:17]: Yes, they are generated to string

Comment (ttkien) [reply lib/src/components/Avatar.re:17]: Here is the code generated

```
// tslint:disable-next-line:interface-over-type-literal
export type size = "small" | "medium" | "large";

```

**Review by toan2406 — DISMISSED** [review]

**Review by toan2406 — APPROVED** [review]

**Review by thai-eh — APPROVED** [review]

**Pull request merged**